### ISSUE TITLE
Simplify names of che.properties to be user friendly

### ISSUE THREAD

**Issue by garagatyi** [eclipse-che/che#2015]
Tyler Jewell:

> I have gone through and done a revision of the default che.properties file. Every property in a comment `#` is the name of the existing property.  The uncommented names are the proposed new property names. I grouped properties by the end user names that we give to concepts as opposed to the internal names, which are not easily understood.

```
#
# Copyright (c) 2012-2016 Codenvy, S.A.
# All rights reserved. This program and the accompanying materials
# are made available under the terms of the Eclipse Public License v1.0
# which accompanies this distribution, and is available at
# http://www.eclipse.org/legal/epl-v10.html
#
# Contributors:
#   Codenvy, S.A. - initial API and implementation
#

### CHE SERVER
# Folder where Che will store internal data objects
che.database=${che.home}/db
#che.conf.storage=${che.home}/storage

# API service. Browsers initiate REST communications to Che server with this URL
che.api=http://localhost:${SERVER_PORT}/wsmaster/api
#api.endpoint=http://localhost:${SERVER_PORT}/wsmaster/api

### WORKSPACES
# The location where your workspaces (and their projects) are stored by Che server.
# This property works for Docker for Windows, Mac, and Linux.
# On vbox for windows, this property is ignored and workspaces stored:
# c:\%userprofile%\AppData\Local\Eclipse Che\
che.workspace.storage=${che.home}/workspaces
che.workspace.storage.create_folders=true

# Your projects are synchronized from the Che server into the machine running each
# workspace. This is the directory in the ws runtime where your projects are mounted.
che.workspace.projects.storage=/projects
# che.machine.projects.internal.storage=/projects

# Configures proxies used by runtimes powering workspaces
che.workspace.http_proxy=
che.workspace.https_proxy=
che.workspace.no_proxy=
#http.proxy=
#https.proxy=
#no_proxy=

# During the stop of the workspace automatically creates a snapshot if the value is {true},
# Otherwise just stops the workspace.
che.workspace.auto_snapshot=true
# workspace.runtime.auto_snapshot=true

# During the start of the workspace automatically restore it from a snapshot if the value is {true},
# Otherwise create a new workspace.
che.workspace.auto_restore=true
# workspace.runtime.auto_restore=true

# Java command line options used to start Che agent in workspace runtime
che.workspace.java_opts=-Xms256m -Xmx2048m -Djava.security.egd=file:/dev/./urandom
# che.machine.java_opts=-Xms256m -Xmx2048m -Djava.security.egd=file:/dev/./urandom

# Folder that will be a volume mount from your host into your workspace.
che.workspace.volume=
# machine.server.extra.volume

# If your workspace creates SSH connections, their timeouts are managed by this property.
che.workspace.ssh_connection_timeout_ms=3000
# machine.ssh.connection_timeout_ms=3000

# The location of the Web terminal used within the browser.
# This is copied into the workspace and runs within the workspace.
# Suffix helps differentiate archive for different architectures/OSes
che.workspace.terminal_linux_amd64=${che.home}/lib/linux_amd64/terminal
che.workspace.terminal_linux_arm7=${che.home}/lib/linux_arm7/terminal
#machine.server.terminal.path_to_archive.linux_amd64=${che.home}/lib/linux_amd64/terminal
#machine.server.terminal.path_to_archive.linux_arm7=${che.home}/lib/linux_arm7/terminal

# Folder where the workspace will store logs from agents and other runtimes
che.workspace.logs=${che.logs.dir}/machine/logs
#machine.logs.location=${che.logs.dir}/machine/logs

# RAM default for new machines.
# TODO: is this per machine, or for the entire environment?
che.workspace.default_memory_mb=1024
#machine.default_mem_size_mb=1024

# This archive contains the server to run the workspace agent and any custom extensions.
# Che injects this archive into machines when they are booted or started.
# Each machine can have many agents installed, each located in a different location.
che.workspace.agent["dev"]=${che.home}/lib/ws-agent.zip
#machine.server.ext.archive=${che.home}/lib/ws-agent.zip



# When the Che server launches a new workspace, Che pings a mini Che server running inside of the
# workspace runtime. We call this mini-Che an "agent". The Che server knows that the workspace
# is ready for user when the agent returns a valid response.
che.workspace.agent["dev"].max_start_time_ms=60000
che.workspace.agent["dev"].ping_delay_ms=2000
che.workspace.agent["dev"].ping_conn_timeout_ms=2000
che.workspace.agent["dev"].ping_timeout_error_msg=Timeout. The Che server is unable to ping your workspace. This implies a network configuration issue, workspace boot failure, or an unusually slow workspace boot.
#machine.ws_agent.max_start_time_ms=60000
#machine.ws_agent.ping_delay_ms=2000
#machine.ws_agent.ping_conn_timeout_ms=2000
#machine.ws_agent.ping_timed_out_error_msg=Timeout reached. The Che server has been unable to verify that your workspace's agent has successfully booted. Either the workspace is unreachable, the agent had an error during startup, or your workspace is starting slowly. You can configure machine.ws_agent.max_start_time_ms in Che properties to increase the timeout.

# Hosts listed here will be added to /etc/hosts of each workspace.
che.workspace.hosts=NULL
#machine.docker.machine_extra_hosts=NULL

# This is the API endpoint of the workspace master running within the core Che server.
# This tells the workspace agent how to connect back to the Che server.
# che-host is a hostname entry added to /etc/hosts of the workspace by the Che server.
che.workspace.che_server_endpoint=http://che-host:${SERVER_PORT}/wsmaster/api
#machine.docker.che_api.endpoint=http://che-host:${SERVER_PORT}/wsmaster/api


### TEMPLATES
# Folder that contains JSON files with code templates and samples
che.template.storage=${che.home}/templates
# project.template_description.location_dir=${che.home}/templates


### STACKS
# File name containing default stacks definitions
che.stacks.storage=${che.home}/stacks/stacks.json
# che.stacks.default=${che.home}/stacks/stacks.json

# Folder name where stack images are stored
che.stacks.images=${che.home}/stacks/images
# che.stacks.images.storage=${che.home}/stacks/images

### AUTHENTICATION PARAMETERS
# Che has a single identity implementation, so this does not change the user experience.
# If true, enables user creation at API level
che.auth.user_self_creation=false
# user.self.creation.allowed=false

che.auth.access_denied_error_page=/error-oauth
#auth.oauth.access_denied_error_page=/error-oauth

# Reserved user names
che.auth.reserved_user_names=
# user.reserved_names=

# You can setup GitHub oAuth to automate authentication to remote repositories.
# You need to first register this application with GitHub oAuth.
che.oauth.github.clientid=NULL
che.oauth.github.clientsecret=NULL
che.oauth.github.authuri= https://github.com/login/oauth/authorize
che.oauth.github.tokenuri= https://github.com/login/oauth/access_token
che.oauth.github.redirecturis= http://localhost:${SERVER_PORT}/wsmaster/api/oauth/callback
#oauth.github.clientid=NULL
#oauth.github.clientsecret=NULL
#oauth.github.authuri= https://github.com/login/oauth/authorize
#oauth.github.tokenuri= https://github.com/login/oauth/access_token
#oauth.github.redirecturis= http://localhost:${SERVER_PORT}/wsmaster/api/oauth/callback


### DOCKER PARAMETERS
# Docker is the default machine implementation within Che. Workspaces are powered by machines
# that are constructed when the workspace is started. The images used to generate containers
# for the machines can come from DockerHub or a private Docker registry.
che.docker.registry=${CHE_REGISTRY_HOST}:5000
#machine.docker.registry=${CHE_REGISTRY_HOST}:5000

# If this is true, then we always pull an image from a registry even if we have an image cached
# locally. If false, Docker only pulls image if it does not exist locally.
che.docker.always_pull_image=true
#machine.docker.pull_image=true

# If true, then launches all workspace runtime containers with Docker's
# privilged mode. Please use responsibly. This is required if you want Che workspaces
# to be able to launch their own runtimes which are embedded Docker containers.
che.docker.privilege=false
#machine.docker.privilege_mode=false

# If the browser clients that are accessing Che are remote AND the configuration of Docker is an
# internal IP address or using Unix sockets, then remote browser clients will not be able to connect
# to the workspace. Set the Docker configuration so that Docker containers have an external IP 
# address and provide that external host or IP address here. 
# This is uncommon, and only needed for advanced Docker configurations.
che.docker.ip=NULL
#machine.docker.local_node_host=NULL

# Use a Docker registry for workspace snapshots. If false, snaps are saved to disk.
che.docker.registry_for_snapshots=false
#machine.docker.snapshot_use_registry=false

# Registry snapshot namespace
che.docker.namespace=NULL
#machine.docker.snapshot.registry_namespace=NULL

# Docker unused containers cleanup period
che.docker.unused_containers_cleanup_min=60
#machine.docker.unused_containers_cleanup_period_min=60

# Version number of the Docker API used within the Che implementation
che.docker.api=1.20
#docker.api.version=1.20

che.docker.tcp_connection_timeout_ms=600000
#docker.connection.tcp.connection_timeout_ms=600000
che.docker.tcp_connection_read_timeout_ms=600000
#docker.connection.tcp.read_timeout_ms=600000

# Docker registry example. Uncomment to add a registry configuration.
# You can configure multiple registries with different names.
#docker.registry.auth.<insert-name>.url=https://index.docker.io/v1/
#docker.registry.auth.<insert-name>.username=<username>
#docker.registry.auth.<insert-name>.password=<password>

# Allows to adjust machine swap memory by multiplying current machnine memory to provided value.
# default is -1 which is unlimited swap. If set, value is multipled by machine memory set by user 
# to determine swap size. To disable swap set to 0.
che.docker.swap=-1
#machine.docker.memory_swap_multiplier=-1


### INTERNAL
# Remove locations where internal message bus events should be propagated to.
# For debugging - set to retrieve internal events from external clients.
notification.server.propagate_events=

# Che extensions can be scheduled executions on a time basis. 
# This configures the size of the thread pool allocated to extensions that are launched on
# a recurring schedule.
schedule.core_pool_size=10

# Everrest is a Java Web Services toolkit that manages JAX-RS & web socket communications
# Users should rarely need to configure this.
# Disable asynchronous mechanism that is embedded in everrest.
org.everrest.asynchronous=false
# Quantity of asynchronous requests which may be processed at the same time
org.everrest.asynchronous.pool.size=20
# Size of queue. If asynchronous request can't be processed after consuming it will be added in queue.
org.everrest.asynchronous.queue.size=500
# Timeout in minutes for request. If after timeout request is not done or client did not come yet to get result of request it may be discarded.
org.everrest.asynchronous.job.timeout=10
# Size of cache for waiting, running and ended request.
org.everrest.asynchronous.cache.size=1024
# Path to asynchronous service
org.everrest.asynchronous.service.path=/async/
```

**Comment by JamesDrummond** [comment]
@garagatyi Can you label this as kind/question or kind/enhancement if you are asking/waiting for @TylerJewell to review?

**Comment by skabashnyuk** [comment]
```
che.workspace.agent["dev"]=${che.home}/lib/ws-agent.zip
#machine.server.ext.archive=${che.home}/lib/ws-agent.zip
```

AFAIK che.workspace.agent["dev"] this syntax is not supported

**Comment by skabashnyuk** [comment]
```
# When the Che server launches a new workspace, Che pings a mini Che server running inside of the
# workspace runtime. We call this mini-Che an "agent". The Che server knows that the workspace
# is ready for user when the agent returns a valid response.
che.workspace.agent["dev"].max_start_time_ms=60000
che.workspace.agent["dev"].ping_delay_ms=2000
che.workspace.agent["dev"].ping_conn_timeout_ms=2000
che.workspace.agent["dev"].ping_timeout_error_msg=Timeout. The Che server is unable to ping your workspace. This implies a network configuration issue, workspace boot failure, or an unusually slow workspace boot.
#machine.ws_agent.max_start_time_ms=60000
#machine.ws_agent.ping_delay_ms=2000
#machine.ws_agent.ping_conn_timeout_ms=2000
#machine.ws_agent.ping_timed_out_error_msg=Timeout reached. The Che server has been unable to verify that your workspace's agent has successfully booted. Either the workspace is unreachable, the agent had an error during startup, or your workspace is starting slowly. You can configure machine.ws_agent.max_start_time_ms in Che properties to increase the timeout.
```

same

**Comment by skabashnyuk** [comment]
Proposed content for aliases file

```
che.database=che.conf.storage
che.api=api.endpoint
che.workspace.projects.storage=che.machine.projects.internal.storage
che.workspace.http_proxy=http.proxy
che.workspace.https_proxy=https.proxy
che.workspace.no_proxy=no_proxy
che.workspace.auto_snapshot=workspace.runtime.auto_snapshot
che.workspace.auto_restore=workspace.runtime.auto_restore
che.workspace.java_opts=che.machine.java_opts
che.workspace.volume=machine.server.extra.volume
che.workspace.ssh_connection_timeout_ms=machine.ssh.connection_timeout_ms
che.workspace.terminal_linux_amd64=machine.server.terminal.path_to_archive.linux_amd64
che.workspace.terminal_linux_arm7=machine.server.terminal.path_to_archive.linux_arm7
che.workspace.logs=machine.logs.location
che.workspace.default_memory_mb=machine.default_mem_size_mb
che.workspace.hosts=machine.docker.machine_extra_hosts
che.workspace.che_server_endpoint=machine.docker.che_api.endpoint
che.workspace.agent.dev=machine.server.ext.archive
che.workspace.agent.dev.max_start_time_ms=machine.ws_agent.max_start_time_ms
che.workspace.agent.dev.ping_delay_ms=machine.ws_agent.ping_delay_ms
che.workspace.agent.dev.ping_conn_timeout_ms=machine.ws_agent.ping_conn_timeout_ms
che.workspace.agent.dev.ping_timeout_error_msg=machine.ws_agent.ping_timed_out_error_msg
che.template.storage=project.template_description.location_dir
che.stacks.storage=che.stacks.default
che.stacks.images=che.stacks.images.storage
che.auth.user_self_creation=user.self.creation.allowed
#TODO remove to default configuration
che.auth.access_denied_error_page=auth.oauth.access_denied_error_page
che.auth.reserved_user_names=user.reserved_names
che.oauth.github.clientid=oauth.github.clientid
che.oauth.github.clientsecret=oauth.github.clientsecret
che.oauth.github.authuri=oauth.github.authuri
che.oauth.github.tokenuri=oauth.github.tokenuri
che.oauth.github.redirecturis=oauth.github.redirecturis
che.docker.registry=machine.docker.registry
che.docker.always_pull_image=machine.docker.pull_image
che.docker.privilege=machine.docker.privilege_mode
che.docker.ip=machine.docker.local_node_host
che.docker.registry_for_snapshots=machine.docker.snapshot_use_registry
che.docker.namespace=machine.docker.snapshot.registry_namespace
che.docker.unused_containers_cleanup_min=machine.docker.unused_containers_cleanup_period_min
che.docker.api=docker.api.version
che.docker.tcp_connection_timeout_ms=docker.connection.tcp.connection_timeout_ms
che.docker.tcp_connection_read_timeout_ms=docker.connection.tcp.read_timeout_m
che.docker.swap=machine.docker.memory_swap_multiplier
```

**Comment by skabashnyuk** [comment]
@TylerJewell  @gazarenkov @garagatyi @evoevodin  @vparfonov please review

**Comment by TylerJewell** [comment]
@skabashnyuk - 
1: I updated to add in the agent parameters with names that have a good format.  If you agree with this, I'll update the default che.properties definition at the beginning of this issue.

2: I found one minor typo where `no_proxy` was missing a letter.  I fixed it in your comment.

3: Can we also have the default `che.properties` file replaced in `/WEB-INF/classes/clodenvy` with the content at the top of this issue?  When we send users to get documentation on the properties that they can configure, it's this file and definitions we want them to see / use.  They will use this as a basis for the names of the environment variables that they can configure.

**Comment by skabashnyuk** [comment]
@TylerJewell @benoitf  can you estimate how this issue may affect on che launcher?

**Comment by benoitf** [comment]
If aliasing is enabled there shouldn't have any issue as it will be backward compliant no ?

**Comment by skabashnyuk** [comment]
It supposed to be so.

**Comment by TylerJewell** [comment]
@skabashnyuk - I have a (probably painful) suggestion.  Can we make the default parameters the new set, and then change the aliasing file to referencing the old names (the current names)?  

By doing this, we will establish that the names that we want to emphasize are the new entries.

@JamesDrummond - we will have to go through docs and slowly update any reference to a che.property to use the new name instead of the old name, but we do not have to do it quickly.

**Comment by skabashnyuk** [comment]
Sorry @TylerJewell  your proposal is not clear to me. Can you rephrase it?

**Comment by benoitf** [comment]
here it how I'm understanding Tyler proposal:
I think it's to have alias file content with the opposite
`che.workspace.volume=machine.server.extra.volume
`==>
`machine.server.extra.volume=che.workspace.volume
`
and in the code we use new name
https://github.com/eclipse/che/blob/d546578366fc06ae7bfd41347d16f260c6e8147f/plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/ExtraVolumeProvider.java#L34

we should inject `che.workspace.volume` (the new name) instead of `machine.server.extra.volume`

**Comment by skabashnyuk** [comment]
Still not clear
As I understand this issue we need.
1. We need to change names of properties in our code to new one.
2. Replace configuration parameters in our properties files with new names
3. Add aliases file to match old user parameters with new parameter names in our code.
Are we on the same page?

**Comment by benoitf** [comment]
@skabashnyuk yes it's as I've understood the last proposal

**Comment by TylerJewell** [comment]
Yes, agree Sergii.  You itemized it correctly.

**Comment by skabashnyuk** [comment]
ok. I see. And we are doing that according to plan listed above.
@TylerJewell your message is still unclear to me.

>  I have a (probably painful) suggestion. Can we make the default parameters the new set, and >then change the aliasing file to referencing the old names (the current names)?
> 
> By doing this, we will establish that the names that we want to emphasize are the new entries.

**Comment by TylerJewell** [comment]
It is this task -  also.  I had created a new Che.properties file with all of the new names to use as a default.  I had organized it for readability and provided some additional documentation for various properties.   Let's place that file into the /conf directory as the default.
1. We need to change names of properties in our code to new one.
2. Replace configuration parameters in our properties files with new names
3. Add aliases file to match old user parameters with new parameter names in our code.
   Are we on the same page?

**Comment by skabashnyuk** [comment]
Right now we are going to put it in `assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/` @TylerJewell  can you clarify what kind of `/conf` do you mean?

**Comment by skabashnyuk** [comment]
FYI. Since we need `che.api` inside of workspace agent code. We will setup environment variable `CHE_CHE_API` for that.

**Comment by benoitf** [comment]
the name of the property looks odd no ? `CHE_CHE_API` instead of `CHE_API`

**Comment by TylerJewell** [comment]
@skabashnyuk - yes, good question.  I was not clear on what I meant before.  The true default is located in `assembly-wsmaster-war/src/main/webapp/WEB-INF/classes`.  However, in the default assembly that is packaged and distributed, I believe a simplified version of this is also located in `/conf` of the assembly itself.  Essentially, hte idea was to only put properties that were essential for users to have to customize and properties that were related to internals (like how websockets are configured) could be removed. Since there was a property precedence built-in, then the che.properties that was in the `/conf` folder was easy to find and easy(er) for new users to understand.  So we should be consistent with this and place a stripped down version into the /conf folder the way we do now.

@skabashnyuk - can you explain why `che.api` needs `CHE_CHE_API` variable but `che.workspace` would only need `CHE_WORKSPACE` variable?

**Comment by garagatyi** [comment]
We have to use CHE_CHE_API variable because our bootstrap module removes CHE_ prefix from injected env variables.

**Comment by benoitf** [comment]
it could be smarter. From user perspective CHE_API for a property named "che.api" is more obvious than "CHE_CHE_API"

**Comment by TylerJewell** [comment]
Yes, agree - we should not impose technology limitations upon the user here.  CHE_CHE_API is not an ok property name.  Let's find a different answer.

**Comment by skabashnyuk** [comment]
Let's be consistent.
@TylerJewell  what should we do with names. Remove `CHE_` or not?

**Comment by TylerJewell** [comment]
What are the choices - I don't know what you are proposing.

**Comment by skabashnyuk** [comment]
1. We do not remove `CHE_` prefix. So `CHE_API=che.api`
2. We do remove `CHE_` prefix. So `CHE_CHE_API=che.api`

**Comment by TylerJewell** [comment]
CHE_API should be equal to che.api.  What are the implications for other property names?

**Comment by skabashnyuk** [comment]
All environment variables for che and codenvy should start from `CHE_`, all properties should start with `che.` too

**Comment by TylerJewell** [comment]
I am not following the implications.  We should have the ability to name any property with any value that we want.  Where did this requirement of all codenvy properties must start with CHE_ come from?

**Comment by skabashnyuk** [comment]
We have a filter for environment variables that is mapped as properties. Variables have to start with `CHE_`

**Comment by TylerJewell** [comment]
-1 - that restriction is too restrictive.

**Comment by mkuznyetsov** [comment]
full che.properties with new names:

```
### CHE SERVER
# Folder where Che will store internal data objects
che.database=${che.home}/storage

# API service. Browsers initiate REST communications to Che server with this URL
che.api=http://localhost:${SERVER_PORT}/wsmaster/api

# Your projects are synchronized from the Che server into the machine running each
# workspace. This is the directory in the ws runtime where your projects are mounted.
che.workspace.storage=${che.home}/workspaces

# Your projects are synchronized from the Che server into the machine running each
# workspace. This is the directory in the machine where your projects are placed.
che.workspace.projects.storage=/projects

# Configures proxies used by runtimes powering workspaces
che.workspace.http_proxy=
che.workspace.https_proxy=
che.workspace.no_proxy=

# During the stop of the workspace automatically creates a snapshot if the value is {true},
# Otherwise just stops the workspace.
che.workspace.auto_snapshot=true

# During the start of the workspace automatically restore it from a snapshot if the value is {true},
# Otherwise create a new workspace.
che.workspace.auto_restore=true

# Java command line options used to start Che agent in workspace runtime
che.workspace.java_opts=-Xms256m -Xmx2048m -Djava.security.egd=file:/dev/./urandom

# Folder that will be a volume mount from your host into your workspace.
che.workspace.volume=

# If your workspace creates SSH connections, their timeouts are managed by this property.
che.workspace.ssh_connection_timeout_ms=10000

# The location of the Web terminal used within the browser.
# This is copied into the workspace and runs within the workspace.
# Suffix helps differentiate archive for different architectures/OSes
che.workspace.terminal_linux_amd64=${che.home}/lib/linux_amd64/terminal
che.workspace.terminal_linux_arm7=${che.home}/lib/linux_arm7/terminal

# Folder where the workspace will store logs from agents and other runtimes
che.workspace.logs=${che.logs.dir}/machine/logs

# RAM default for new machines.
# TODO: is this per machine, or for the entire environment?
che.workspace.default_memory_mb=1024

# This archive contains the server to run the workspace agent and any custom extensions.
# Che injects this archive into machines when they are booted or started.
# Each machine can have many agents installed, each located in a different location.
che.workspace.agent.dev=${che.home}/lib/ws-agent.tar.gz

# When the Che server launches a new workspace, Che pings a mini Che server running inside of the
# workspace runtime. We call this mini-Che an "agent". The Che server knows that the workspace
# is ready for user when the agent returns a valid response.
che.workspace.agent.dev.max_start_time_ms=180000
che.workspace.agent.dev.ping_delay_ms=2000
che.workspace.agent.dev.ping_conn_timeout_ms=2000
che.workspace.agent.dev.ping_timeout_error_msg=Timeout. The Che server is unable to ping your workspace. This implies a network configuration issue, workspace boot failure, or an unusually slow workspace boot.

# Used for other agents like terminal, ssh
che.agent.max_start_time_ms=120000
che.agent.ping_delay_ms=2000

# Hosts listed here will be added to /etc/hosts of each workspace.
che.workspace.hosts=NULL

# This is the API endpoint of the workspace master running within the core Che server.
# This tells the workspace agent how to connect back to the Che server.
# che-host is a hostname entry added to /etc/hosts of the workspace by the Che server.
che.workspace.che_server_endpoint=http://che-host:${SERVER_PORT}/wsmaster/api


### TEMPLATES
# Folder that contains JSON files with code templates and samples
che.template.storage=${che.home}/templates

### STACKS
# File name containing default stacks definitions
che.stacks.storage=${che.home}/stacks/stacks.json

# che.stacks.default=${che.home}/stacks/stacks.json
# Folder name where stack images are stored
che.stacks.images=${che.home}/stacks/images

### AUTHENTICATION PARAMETERS
# Che has a single identity implementation, so this does not change the user experience.
# If true, enables user creation at API level
che.auth.user_self_creation=false

che.auth.access_denied_error_page=/error-oauth

# Reserved user names
che.auth.reserved_user_names=

# You can setup GitHub oAuth to automate authentication to remote repositories.
# You need to first register this application with GitHub oAuth.
che.oauth.github.clientid=NULL
che.oauth.github.clientsecret=NULL
che.oauth.github.authuri= https://github.com/login/oauth/authorize
che.oauth.github.tokenuri= https://github.com/login/oauth/access_token
che.oauth.github.redirecturis= http://localhost:${SERVER_PORT}/wsmaster/api/oauth/callback


### DOCKER PARAMETERS
# Docker is the default machine implementation within Che. Workspaces are powered by machines
# that are constructed when the workspace is started. The images used to generate containers
# for the machines can come from DockerHub or a private Docker registry.
che.docker.registry=${CHE_REGISTRY_HOST}:5000

# If this is true, then we always pull an image from a registry even if we have an image cached
# locally. If false, Docker only pulls image if it does not exist locally.
che.docker.always_pull_image=true

# If true, then launches all workspace runtime containers with Docker's
# privilged mode. Please use responsibly. This is required if you want Che workspaces
# to be able to launch their own runtimes which are embedded Docker containers.
che.docker.privilege=false

# If the browser clients that are accessing Che are remote AND the configuration of Docker is an
# internal IP address or using Unix sockets, then remote browser clients will not be able to connect
# to the workspace. Set the Docker configuration so that Docker containers have an external IP
# address and provide that external host or IP address here.
# This is uncommon, and only needed for advanced Docker configurations.
che.docker.ip=NULL

# The hostname that a browser should use to connect to a workspace container.
# Only set this if your workspace containers are not directly pingable.
# This is unusual, but happens for example in Docker for Mac when containers are in a VM.
che.docker.ip.external=NULL

# Provides a Docker network where Che server is running.
# Workspace containers created by Che will be added to this Docker network.
# Communications between the Che server and container occur over this network.
che.docker.network=NULL

# Use a Docker registry for workspace snapshots. If false, snaps are saved to disk.
che.docker.registry_for_snapshots=false

# Registry snapshot namespace
che.docker.namespace=NULL

# Docker unused containers cleanup period
che.docker.unused_containers_cleanup_min=60

# Version number of the Docker API used within the Che implementation
che.docker.api=1.20


che.docker.tcp_connection_timeout_ms=600000
che.docker.tcp_connection_read_timeout_ms=600000

# Docker registry example. Uncomment to add a registry configuration.
# You can configure multiple registries with different names.
#docker.registry.auth.<insert-name>.url=https://index.docker.io/v1/
#docker.registry.auth.<insert-name>.username=<username>
#docker.registry.auth.<insert-name>.password=<password>

# Allows to adjust machine swap memory by multiplying current machnine memory to provided value.
# default is -1 which is unlimited swap. If set, value is multipled by machine memory set by user
# to determine swap size. To disable swap set to 0.
che.docker.swap=-1
#machine.docker.memory_swap_multiplier=-1


### INTERNAL
# Remove locations where internal message bus events should be propagated to.
# For debugging - set to retrieve internal events from external clients.
notification.server.propagate_events=

# Che extensions can be scheduled executions on a time basis.
# This configures the size of the thread pool allocated to extensions that are launched on
# a recurring schedule.
schedule.core_pool_size=10

# Everrest is a Java Web Services toolkit that manages JAX-RS & web socket communications
# Users should rarely need to configure this.
# Disable asynchronous mechanism that is embedded in everrest.
org.everrest.asynchronous=false
# Quantity of asynchronous requests which may be processed at the same time
org.everrest.asynchronous.pool.size=20
# Size of queue. If asynchronous request can't be processed after consuming it will be added in queue.
org.everrest.asynchronous.queue.size=500
# Timeout in minutes for request. If after timeout request is not done or client did not come yet to get result of request it may be discarded.
org.everrest.asynchronous.job.timeout=10
# Size of cache for waiting, running and ended request.
org.everrest.asynchronous.cache.size=1024
# Path to asynchronous service
org.everrest.asynchronous.service.path=/async/


```

**Comment by TylerJewell** [comment]
@skabashnyuk @mkuznyetsov - this file is not organized the same way that I had the original file orgnaized.  Can you take a look at the original posting file - and organize along the same structure, bringing like-minded things together and then grouping them into sections that reflect how our product is organized.

**Comment by mkuznyetsov** [comment]
@TylerJewell I've updated the comment above with ordered # commentd properties as in original posting file.
There are some properties left, that were not listed there, do you want to rename/reorder them?

```
# If the browser cannot access directly to the IP of the Docker host (e.g. when docker host is behind a NAT or
# with Docker for Mac) we can set this external address to let the browser access Docker containers.
# wsmaster will still use machine.docker.local_node_host to access Docker containers containers.
machine.docker.local_node_host.external=NULL

# Provides docker network where Che server is running.
# This allows to configure communication between Che server and Che workspaces.
# All workspace containers in Che will be added to this network to be able to connect to Che server.
docker.che_host_network=NULL
```

these particluar properties look similar to old "machine.ws-agent.*", except that they are used for other agents (terminal, ssh).

```
# When Che start an agent, it performs check if it is launched.
machine.agent.max_start_time_ms=120000
machine.agent.ping_delay_ms=2000
```

**Comment by TylerJewell** [comment]
@mkuznyetsov 

The first one that you mentioned, `machine.docker.local_node_host.external=NULL` is listed in the first document at the top.  It is:

```
# If the browser clients that are accessing Che are remote AND the configuration of Docker is an
# internal IP address or using Unix sockets, then remote browser clients will not be able to connect
# to the workspace. Set the Docker configuration so that Docker containers have an external IP 
# address and provide that external host or IP address here. 
# This is uncommon, and only needed for advanced Docker configurations.
che.docker.ip=NULL
#machine.docker.local_node_host=NULL
```

The second one, `docker.che_host_network=NULL`, let's add this one as:

```
# Provides a Docker network where Che server is running.
# Workspace containers created by Che will be added to this Docker network.
# Communications between the Che server and container occur over this network.
che.docker.network=NULL
```

For these two:

```
machine.agent.max_start_time_ms=120000
machine.agent.ping_delay_ms=2000
```

they should be in the agent configuration.  They are listed above, but maybe not obvious.  If these properties apply to ALL agents (not just ws-agent), then let's rename name to be clear:

```
# Che workspaces run agents in their runtime. Some of these agents are mini Che servers that 
# can (and must) be pingable from Che. Che confirms a workspace's readiness when an agent
# returns valid responses.
che.workspace.agent.max_start_time_ms=180000
che.workspace.agent.ping_delay_ms=2000
che.workspace.agent.ping_conn_timeout_ms=2000
che.workspace.agent.ping_timeout_error_msg=Timeout. Che is unable to ping agents in your workspace. This is a network issue, boot failure, or an unusually slow boot.
```

**Comment by mkuznyetsov** [comment]
@TylerJewell so it is
machine.docker.local_node_host=che.docker.ip
machine.docker.local_node_host.external=che.docker.ip.external?

**Comment by TylerJewell** [comment]
@mkuznyetsov:

```
machine.docker.local_node_host=che.docker.ip
machine.docker.local_node_host.external=che.docker.network
```

And then make sure you get the appropriate docs for each one to put into the /conf/che.properties file:

```
# If the browser clients that are accessing Che are remote AND the configuration of Docker is an
# internal IP address or using Unix sockets, then remote browser clients will not be able to connect
# to the workspace. Set the Docker configuration so that Docker containers have an external IP 
# address and provide that external host or IP address here. 
# This is uncommon, and only needed for advanced Docker configurations.
che.docker.ip=NULL

# Provides a Docker network where Che server is running.
# Workspace containers created by Che will be added to this Docker network.
# Communications between the Che server and container occur over this network.
che.docker.network=NULL
```

**Comment by mkuznyetsov** [comment]
@TylerJewell all properties are updated as agreed in https://github.com/eclipse/che/issues/2015#issuecomment-253237734

**Comment by TylerJewell** [comment]
@mkuznyetsov @skabashnyuk what is the merge plan for this?  Are we wanting to do this before or after M6 is released?  I suspect that we have a few third party solutions that will need to be aggressively tested as soon as this is merged, particularly @benoitf chedir, @tylerjewell che-launcher, and the new Dockerfile for che-server.
